File: telemetry/aitf/sdk/typescript/src/instrumentation/agent.ts
/**
 * AITF Agent Instrumentation.
 *
 * Provides tracing for AI agent operations: sessions, steps, delegation,
 * multi-agent orchestration, and memory access.
 */

import {
  trace,
  Span,
  SpanKind,
  SpanStatusCode,
  Tracer,
  TracerProvider,
} from "@opentelemetry/api";
import { randomUUID } from "crypto";
import {
  AgentAttributes,
  MemoryAttributes,
} from "../semantic-conventions/attributes";

const TRACER_NAME = "instrumentation.agent";

/** Options for tracing an agent session. */
export interface TraceSessionOptions {
  agentName: string;
  agentId?: string;
  agentType?: string;
  framework?: string;
  version?: string;
  description?: string;
  sessionId?: string;
  teamName?: string;
}

/** Options for tracing a multi-agent team. */
export interface TraceTeamOptions {
  teamName: string;
  teamId?: string;
  topology?: string;
  members?: string[];
  coordinator?: string;
}

/**
 * Helper class for setting attributes on an agent step span.
 */
export class AgentStep {
  private readonly _span: Span;

  constructor(span: Span) {
    this._span = span;
  }

  get span(): Span {
    return this._span;
  }

  setThought(thought: string): void {
    this._span.setAttribute(AgentAttributes.STEP_THOUGHT, thought);
  }

  setAction(action: string): void {
    this._span.setAttribute(AgentAttributes.STEP_ACTION, action);
  }

  setObservation(observation: string): void {
    this._span.setAttribute(AgentAttributes.STEP_OBSERVATION, observation);
  }

  setStatus(status: string): void {
    this._span.setAttribute(AgentAttributes.STEP_STATUS, status);
  }
}

/**
 * Helper for managing an agent session's spans.
 */
export class AgentSession {
  private readonly _span: Span;
  private readonly _tracer: Tracer;
  private readonly _agentName: string;
  private readonly _sessionId: string;
  private _stepCount = 0;

  constructor(span: Span, tracer: Tracer, agentName: string, sessionId: string) {
    this._span = span;
    this._tracer = tracer;
    this._agentName = agentName;
    this._sessionId = sessionId;
  }

  get span(): Span {
    return this._span;
  }

  get stepCount(): number {
    return this._stepCount;
  }

  /**
   * Create a child span for an agent step.
   */
  step<T>(stepType: string, fn: (step: AgentStep) => T): T;
  step(stepType: string): AgentStep;
  step<T>(
    stepType: string,
    fn?: (step: AgentStep) => T
  ): AgentStep | T {
    this._stepCount++;
    const attributes: Record<string, string | number | boolean> = {
      [AgentAttributes.NAME]: this._agentName,
      [AgentAttributes.STEP_TYPE]: stepType,
      [AgentAttributes.STEP_INDEX]: this._stepCount,
    };

    if (fn) {
      return this._tracer.startActiveSpan(
        `agent.step.${stepType} ${this._agentName}`,
        { kind: SpanKind.INTERNAL, attributes },
        (otelSpan) => {
          const agentStep = new AgentStep(otelSpan);
          try {
            const result = fn(agentStep);
            if (result instanceof Promise) {
              return (result as Promise<unknown>)
                .then((val) => {
                  otelSpan.setStatus({ code: SpanStatusCode.OK });
                  otelSpan.end();
                  return val;
                })
                .catch((err) => {
                  otelSpan.setStatus({
                    code: SpanStatusCode.ERROR,
                    message: String(err),
                  });
                  otelSpan.recordException(err);
                  otelSpan.end();
                  throw err;
                }) as T;
            }
            otelSpan.setStatus({ code: SpanStatusCode.OK });
            otelSpan.end();
            return result;
          } catch (err) {
            otelSpan.setStatus({
              code: SpanStatusCode.ERROR,
              message: String(err),
            });
            otelSpan.recordException(err as Error);
            otelSpan.end();
            throw err;
          }
        }
      );
    }

    const otelSpan = this._tracer.startSpan(
      `agent.step.${stepType} ${this._agentName}`,
      { kind: SpanKind.INTERNAL, attributes }
    );
    return new AgentStep(otelSpan);
  }

  /**
   * Create a delegation span.
   */
  delegate<T>(
    targetAgent: string,
    options: {
      targetAgentId?: string;
      reason?: string;
      strategy?: string;
      task?: string;
    },
    fn: (span: Span) => T
  ): T {
    this._stepCount++;
    const targetAgentId = options.targetAgentId ?? randomUUID();
    const attributes: Record<string, string | number | boolean> = {
      [AgentAttributes.NAME]: this._agentName,
      [AgentAttributes.STEP_TYPE]: AgentAttributes.StepType.DELEGATION,
      [AgentAttributes.STEP_INDEX]: this._stepCount,
      [AgentAttributes.DELEGATION_TARGET_AGENT]: targetAgent,
      [AgentAttributes.DELEGATION_TARGET_AGENT_ID]: targetAgentId,
      [AgentAttributes.DELEGATION_STRATEGY]: options.strategy ?? "capability",
    };
    if (options.reason) {
      attributes[AgentAttributes.DELEGATION_REASON] = options.reason;
    }
    if (options.task) {
      attributes[AgentAttributes.DELEGATION_TASK] = options.task;
    }

    return this._tracer.startActiveSpan(
      `agent.delegate ${this._agentName} -> ${targetAgent}`,
      { kind: SpanKind.INTERNAL, attributes },
      (otelSpan) => {
        try {
          const result = fn(otelSpan);
          if (result instanceof Promise) {
            return (result as Promise<unknown>)
              .then((val) => {
                otelSpan.setStatus({ code: SpanStatusCode.OK });
                otelSpan.end();
                return val;
              })
              .catch((err) => {
                otelSpan.setStatus({
                  code: SpanStatusCode.ERROR,
                  message: String(err),
                });
                otelSpan.recordException(err);
                otelSpan.end();
                throw err;
              }) as T;
          }
          otelSpan.setStatus({ code: SpanStatusCode.OK });
          otelSpan.end();
          return result;
        } catch (err) {
          otelSpan.setStatus({
            code: SpanStatusCode.ERROR,
            message: String(err),
          });
          otelSpan.recordException(err as Error);
          otelSpan.end();
          throw err;
        }
      }
    );
  }

  /**
   * Create a memory access span.
   */
  memoryAccess<T>(
    operation: string,
    options: { store?: string; key?: string },
    fn: (span: Span) => T
  ): T {
    const attributes: Record<string, string | number | boolean> = {
      [AgentAttributes.NAME]: this._agentName,
      [MemoryAttributes.OPERATION]: operation,
      [MemoryAttributes.STORE]: options.store ?? "short_term",
    };
    if (options.key) {
      attributes[MemoryAttributes.KEY] = options.key;
    }

    return this._tracer.startActiveSpan(
      `agent.memory.${operation} ${this._agentName}`,
      { kind: SpanKind.INTERNAL, attributes },
      (otelSpan) => {
        try {
          const result = fn(otelSpan);
          if (result instanceof Promise) {
            return (result as Promise<unknown>)
              .then((val) => {
                otelSpan.setStatus({ code: SpanStatusCode.OK });
                otelSpan.end();
                return val;
              })
              .catch((err) => {
                otelSpan.setStatus({
                  code: SpanStatusCode.ERROR,
                  message: String(err),
                });
                otelSpan.recordException(err);
                otelSpan.end();
                throw err;
              }) as T;
          }
          otelSpan.setStatus({ code: SpanStatusCode.OK });
          otelSpan.end();
          return result;
        } catch (err) {
          otelSpan.setStatus({
            code: SpanStatusCode.ERROR,
            message: String(err),
          });
          otelSpan.recordException(err as Error);
          otelSpan.end();
          throw err;
        }
      }
    );
  }

  /** Finalize the session span with turn count and OK status. */
  end(): void {
    this._span.setAttribute(
      AgentAttributes.SESSION_TURN_COUNT,
      this._stepCount
    );
    this._span.setStatus({ code: SpanStatusCode.OK });
    this._span.end();
  }

  /** End the session span with an error. */
  endWithError(err: Error): void {
    this._span.setStatus({
      code: SpanStatusCode.ERROR,
      message: String(err),
    });
    this._span.recordException(err);
    this._span.end();
  }
}

/**
 * Instrumentor for AI agent operations.
 */
export class AgentInstrumentor {
  private _tracerProvider: TracerProvider | null;
  private _tracer: Tracer | null = null;
  private _instrumented = false;

  constructor(tracerProvider?: TracerProvider) {
    this._tracerProvider = tracerProvider ?? null;
  }

  instrument(): void {
    const tp = this._tracerProvider ?? trace.getTracerProvider();
    this._tracer = tp.getTracer(TRACER_NAME);
    this._instrumented = true;
  }

  uninstrument(): void {
    this._tracer = null;
    this._instrumented = false;
  }

  getTracer(): Tracer {
    if (!this._tracer) {
      const tp = this._tracerProvider ?? trace.getTracerProvider();
      this._tracer = tp.getTracer(TRACER_NAME);
    }
    return this._tracer;
  }

  /**
   * Trace an agent session.
   *
   * With callback:
   *   agent.traceSession({ agentName: "research-agent" }, async (session) => {
   *     session.step("planning", (step) => {
   *       step.setThought("I need to search for information");
   *     });
   *   });
   *
   * Without callback (manual management):
   *   const session = agent.traceSession({ agentName: "research-agent" });
   *   // ... use session ...
   *   session.end();
   */
  traceSession(options: TraceSessionOptions): AgentSession;
  traceSession<T>(
    options: TraceSessionOptions,
    fn: (session: AgentSession) => T
  ): T;
  traceSession<T>(
    options: TraceSessionOptions,
    fn?: (session: AgentSession) => T
  ): AgentSession | T {
    const tracer = this.getTracer();
    const sessionId = options.sessionId ?? randomUUID();
    const agentId = options.agentId ?? randomUUID();

    const attributes: Record<string, string | number | boolean> = {
      [AgentAttributes.NAME]: options.agentName,
      [AgentAttributes.ID]: agentId,
      [AgentAttributes.TYPE]: options.agentType ?? "autonomous",
      [AgentAttributes.FRAMEWORK]: options.framework ?? "custom",
      [AgentAttributes.SESSION_ID]: sessionId,
    };
    if (options.version) {
      attributes[AgentAttributes.VERSION] = options.version;
    }
    if (options.description) {
      attributes[AgentAttributes.DESCRIPTION] = options.description;
    }
    if (options.teamName) {
      attributes[AgentAttributes.TEAM_NAME] = options.teamName;
    }

    if (fn) {
      return tracer.startActiveSpan(
        `agent.session ${options.agentName}`,
        { kind: SpanKind.INTERNAL, attributes },
        (otelSpan) => {
          const session = new AgentSession(
            otelSpan,
            tracer,
            options.agentName,
            sessionId
          );
          try {
            const result = fn(session);
            if (result instanceof Promise) {
              return (result as Promise<unknown>)
                .then((val) => {
                  session.end();
                  return val;
                })
                .catch((err) => {
                  session.endWithError(err);
                  throw err;
                }) as T;
            }
            session.end();
            return result;
          } catch (err) {
            session.endWithError(err as Error);
            throw err;
          }
        }
      );
    }

    const otelSpan = tracer.startSpan(
      `agent.session ${options.agentName}`,
      { kind: SpanKind.INTERNAL, attributes }
    );
    return new AgentSession(otelSpan, tracer, options.agentName, sessionId);
  }

  /**
   * Trace a multi-agent team orchestration.
   */
  traceTeam<T>(options: TraceTeamOptions, fn: (span: Span) => T): T {
    const tracer = this.getTracer();
    const teamId = options.teamId ?? randomUUID();
    const attributes: Record<string, string | number | boolean | string[]> = {
      [AgentAttributes.TEAM_NAME]: options.teamName,
      [AgentAttributes.TEAM_ID]: teamId,
      [AgentAttributes.TEAM_TOPOLOGY]: options.topology ?? "hierarchical",
    };
    if (options.members) {
      attributes[AgentAttributes.TEAM_MEMBERS] = options.members;
    }
    if (options.coordinator) {
      attributes[AgentAttributes.TEAM_COORDINATOR] = options.coordinator;
    }

    return tracer.startActiveSpan(
      `agent.team.orchestrate ${options.teamName}`,
      { kind: SpanKind.INTERNAL, attributes },
      (otelSpan) => {
        try {
          const result = fn(otelSpan);
          if (result instanceof Promise) {
            return (result as Promise<unknown>)
              .then((val) => {
                otelSpan.setStatus({ code: SpanStatusCode.OK });
                otelSpan.end();
                return val;
              })
              .catch((err) => {
                otelSpan.setStatus({
                  code: SpanStatusCode.ERROR,
                  message: String(err),
                });
                otelSpan.recordException(err);
                otelSpan.end();
                throw err;
              }) as T;
          }
          otelSpan.setStatus({ code: SpanStatusCode.OK });
          otelSpan.end();
          return result;
        } catch (err) {
          otelSpan.setStatus({
            code: SpanStatusCode.ERROR,
            message: String(err),
          });
          otelSpan.recordException(err as Error);
          otelSpan.end();
          throw err;
        }
      }
    );
  }
}

File: telemetry/aitf/sdk/python/src/aitf/instrumentation/agent.py
"""AITF Agent Instrumentation.

Provides tracing for AI agent operations: sessions, steps, delegation,
multi-agent orchestration, and memory access.
"""

from __future__ import annotations

import uuid
from contextlib import contextmanager
from functools import wraps
from typing import Any, Callable, Generator

from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.trace import SpanKind, StatusCode

from aitf.semantic_conventions.attributes import AgentAttributes, GenAIAttributes, MemoryAttributes

_TRACER_NAME = "instrumentation.agent"


class AgentInstrumentor:
    """Instrumentor for AI agent operations."""

    def __init__(self, tracer_provider: TracerProvider | None = None):
        self._tracer_provider = tracer_provider
        self._tracer: trace.Tracer | None = None
        self._instrumented = False

    def instrument(self) -> None:
        tp = self._tracer_provider or trace.get_tracer_provider()
        self._tracer = tp.get_tracer(_TRACER_NAME)
        self._instrumented = True

    def uninstrument(self) -> None:
        self._tracer = None
        self._instrumented = False

    def get_tracer(self) -> trace.Tracer:
        if self._tracer is None:
            tp = self._tracer_provider or trace.get_tracer_provider()
            self._tracer = tp.get_tracer(_TRACER_NAME)
        return self._tracer

    @contextmanager
    def trace_session(
        self,
        agent_name: str,
        agent_id: str | None = None,
        agent_type: str = "autonomous",
        framework: str = "custom",
        version: str | None = None,
        description: str | None = None,
        session_id: str | None = None,
        team_name: str | None = None,
        workflow_id: str | None = None,
        state: str | None = None,
    ) -> Generator[AgentSession, None, None]:
        """Context manager for tracing an agent session.

        Usage:
            with agent.trace_session("research-agent", framework="langchain") as session:
                with session.step("planning") as step:
                    step.set_thought("I need to search for information")
                with session.step("tool_use") as step:
                    result = call_tool(...)
                    step.set_observation(result)
        """
        tracer = self.get_tracer()
        session_id = session_id or str(uuid.uuid4())
        agent_id = agent_id or str(uuid.uuid4())

        attributes: dict[str, Any] = {
            GenAIAttributes.AGENT_NAME: agent_name,
            GenAIAttributes.AGENT_ID: agent_id,
            AgentAttributes.TYPE: agent_type,
            AgentAttributes.FRAMEWORK: framework,
            GenAIAttributes.CONVERSATION_ID: session_id,
        }
        if version:
            attributes[GenAIAttributes.AGENT_VERSION] = version
        if description:
            attributes[GenAIAttributes.AGENT_DESCRIPTION] = description
        if team_name:
            attributes[AgentAttributes.TEAM_NAME] = team_name
        if workflow_id:
            attributes[AgentAttributes.WORKFLOW_ID] = workflow_id
        if state:
            attributes[AgentAttributes.STATE] = state

        with tracer.start_as_current_span(
            name=f"agent.session {agent_name}",
            kind=SpanKind.INTERNAL,
            attributes=attributes,
        ) as span:
            session = AgentSession(span, tracer, agent_name, session_id)
            try:
                yield session
                span.set_attribute(AgentAttributes.SESSION_TURN_COUNT, session.step_count)
                span.set_status(StatusCode.OK)
            except Exception as exc:
                span.set_status(StatusCode.ERROR, str(exc))
                span.record_exception(exc)
                raise

    @contextmanager
    def trace_team(
        self,
        team_name: str,
        team_id: str | None = None,
        topology: str = "hierarchical",
        members: list[str] | None = None,
        coordinator: str | None = None,
    ) -> Generator[trace.Span, None, None]:
        """Context manager for tracing a multi-agent team orchestration."""
        tracer = self.get_tracer()
        team_id = team_id or str(uuid.uuid4())
        attributes: dict[str, Any] = {
            AgentAttributes.TEAM_NAME: team_name,
            AgentAttributes.TEAM_ID: team_id,
            AgentAttributes.TEAM_TOPOLOGY: topology,
        }
        if members:
            attributes[AgentAttributes.TEAM_MEMBERS] = members
        if coordinator:
            attributes[AgentAttributes.TEAM_COORDINATOR] = coordinator

        with tracer.start_as_current_span(
            name=f"agent.team.orchestrate {team_name}",
            kind=SpanKind.INTERNAL,
            attributes=attributes,
        ) as span:
            try:
                yield span
                span.set_status(StatusCode.OK)
            except Exception as exc:
                span.set_status(StatusCode.ERROR, str(exc))
                span.record_exception(exc)
                raise


class AgentSession:
    """Helper for managing an agent session's spans."""

    def __init__(
        self,
        span: trace.Span,
        tracer: trace.Tracer,
        agent_name: str,
        session_id: str,
    ):
        self._span = span
        self._tracer = tracer
        self._agent_name = agent_name
        self._session_id = session_id
        self._step_count = 0

    @property
    def span(self) -> trace.Span:
        return self._span

    @property
    def step_count(self) -> int:
        return self._step_count

    def set_state(self, state: str) -> None:
        """Set the agent lifecycle state (CoSAI WS2: AGENT_TRACE)."""
        self._span.set_attribute(AgentAttributes.STATE, state)

    def set_workflow_id(self, workflow_id: str) -> None:
        """Set the workflow ID (CoSAI WS2: AGENT_TRACE)."""
        self._span.set_attribute(AgentAttributes.WORKFLOW_ID, workflow_id)

    @contextmanager
    def step(
        self,
        step_type: str,
        **kwargs: Any,
    ) -> Generator[AgentStep, None, None]:
        """Create a child span for an agent step."""
        self._step_count += 1
        attributes: dict[str, Any] = {
            GenAIAttributes.AGENT_NAME: self._agent_name,
            AgentAttributes.STEP_TYPE: step_type,
            AgentAttributes.STEP_INDEX: self._step_count,
        }
        # Validate extra attributes to prevent namespace pollution
        _ALLOWED_VALUE_TYPES = (str, int, float, bool)
        _MAX_ATTR_KEY_LEN = 128
        for key, value in kwargs.items():
            if len(key) > _MAX_ATTR_KEY_LEN:
                continue
            if not isinstance(value, _ALLOWED_VALUE_TYPES):
                continue
            attr_key = f"agent.step.{key}"
            attributes[attr_key] = value

        with self._tracer.start_as_current_span(
            name=f"agent.step.{step_type} {self._agent_name}",
            kind=SpanKind.INTERNAL,
            attributes=attributes,
        ) as span:
            step = AgentStep(span)
            try:
                yield step
                span.set_status(StatusCode.OK)
            except Exception as exc:
                span.set_status(StatusCode.ERROR, str(exc))
                span.record_exception(exc)
                raise

    @contextmanager
    def delegate(
        self,
        target_agent: str,
        target_agent_id: str | None = None,
        reason: str | None = None,
        strategy: str = "capability",
        task: str | None = None,
    ) -> Generator[trace.Span, None, None]:
        """Create a delegation span."""
        self._step_count += 1
        target_agent_id = target_agent_id or str(uuid.uuid4())
        attributes: dict[str, Any] = {
            GenAIAttributes.AGENT_NAME: self._agent_name,
            AgentAttributes.STEP_TYPE: AgentAttributes.StepType.DELEGATION,
            AgentAttributes.STEP_INDEX: self._step_count,
            AgentAttributes.DELEGATION_TARGET_AGENT: target_agent,
            AgentAttributes.DELEGATION_TARGET_AGENT_ID: target_agent_id,
            AgentAttributes.DELEGATION_STRATEGY: strategy,
        }
        if reason:
            attributes[AgentAttributes.DELEGATION_REASON] = reason
        if task:
            attributes[AgentAttributes.DELEGATION_TASK] = task

        with self._tracer.start_as_current_span(
            name=f"agent.delegate {self._agent_name} -> {target_agent}",
            kind=SpanKind.INTERNAL,
            attributes=attributes,
        ) as span:
            try:
                yield span
                span.set_status(StatusCode.OK)
            except Exception as exc:
                span.set_status(StatusCode.ERROR, str(exc))
                span.record_exception(exc)
                raise

    @contextmanager
    def memory_access(
        self,
        operation: str,
        store: str = "short_term",
        key: str | None = None,
    ) -> Generator[trace.Span, None, None]:
        """Create a memory access span."""
        attributes: dict[str, Any] = {
            GenAIAttributes.AGENT_NAME: self._agent_name,
            MemoryAttributes.OPERATION: operation,
            MemoryAttributes.STORE: store,
        }
        if key:
            attributes[MemoryAttributes.KEY] = key

        with self._tracer.start_as_current_span(
            name=f"agent.memory.{operation} {self._agent_name}",
            kind=SpanKind.INTERNAL,
            attributes=attributes,
        ) as span:
            try:
                yield span
                span.set_status(StatusCode.OK)
            except Exception as exc:
                span.set_status(StatusCode.ERROR, str(exc))
                span.record_exception(exc)
                raise


class AgentStep:
    """Helper for setting attributes on an agent step span."""

    def __init__(self, span: trace.Span):
        self._span = span

    @property
    def span(self) -> trace.Span:
        return self._span

    def set_thought(self, thought: str) -> None:
        self._span.set_attribute(AgentAttributes.STEP_THOUGHT, thought)

    def set_action(self, action: str) -> None:
        self._span.set_attribute(AgentAttributes.STEP_ACTION, action)

    def set_observation(self, observation: str) -> None:
        self._span.set_attribute(AgentAttributes.STEP_OBSERVATION, observation)

    def set_status(self, status: str) -> None:
        self._span.set_attribute(AgentAttributes.STEP_STATUS, status)

    def set_scratchpad(self, scratchpad: str) -> None:
        """Set the agent scratchpad / working memory (CoSAI WS2: AGENT_TRACE)."""
        self._span.set_attribute(AgentAttributes.SCRATCHPAD, scratchpad)

    def set_next_action(self, next_action: str) -> None:
        """Set the next planned action (CoSAI WS2: AGENT_TRACE)."""
        self._span.set_attribute(AgentAttributes.NEXT_ACTION, next_action)


def agent_span(
    agent_name: str,
    framework: str = "custom",
    agent_type: str = "autonomous",
) -> Callable:
    """Decorator for tracing agent functions.

    Usage:
        @agent_span(agent_name="research-agent", framework="langchain")
        async def research_agent(query: str):
            ...
    """

    def decorator(func: Callable) -> Callable:
        @wraps(func)
        def wrapper(*args: Any, **kwargs: Any) -> Any:
            instrumentor = AgentInstrumentor()
            with instrumentor.trace_session(
                agent_name=agent_name,
                framework=framework,
                agent_type=agent_type,
            ):
                return func(*args, **kwargs)

        @wraps(func)
        async def async_wrapper(*args: Any, **kwargs: Any) -> Any:
            instrumentor = AgentInstrumentor()
            with instrumentor.trace_session(
                agent_name=agent_name,
                framework=framework,
                agent_type=agent_type,
            ):
                return await func(*args, **kwargs)

        import asyncio

        if asyncio.iscoroutinefunction(func):
            return async_wrapper
        return wrapper

    return decorator
